net, net_processing: additional and consistent disconnect logging

Sjors opened the pull request:
While debugging unexpected disconnections, possibly related to #28331, I found some additional [net] logging to be useful.

All cases where we disconnect now come with a log message that has the word `disconnecting`:

* all calls to `CloseSocketDisconnect()` log `disconnecting peer=…`
* wherever we set `pnode->fDisconnect = true;`
* for all `InactivityCheck` cases (which in turn sets `fDisconnect`)
* replaces "dropping" with "disconnecting" in `Network not active, dropping peer=…`

A few exceptions are listed here: https://github.com/bitcoin/bitcoin/pull/28521#discussion_r1890824361

I changed `CloseSocketDisconnect()` to no longer log `disconnecting`, and instead have all the call sites do so.

This PR introduces two helper functions on `CNode`: `DisconnectMsg` and `LogIP`. The second and third commit use these helpers in `net_processing.cpp` so these disconnect messages are more consistent now (e.g. some didn't log the IP). No new messages are added there though.

The `LogIP()` helper is rarely used outside of a disconnect event, but it's available for future use.

Any `LogPrint` this PR touches is replaced with `LogDebug` (superseded by #30750), and every `LogPrintf ` with `LogInfo`.

ajtowns commented:
Concept ACK -- always makes sense to debug log disconnection reasons IMO

0xB10C commented:
ACK 320978662db6b17e61afbf77a6584e33181c12b9

ajtowns commented on src/net.cpp:
```diff
@@ -2138,7 +2139,10 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
 
-        if (InactivityCheck(*pnode)) pnode->fDisconnect = true;
+        if (InactivityCheck(*pnode)) {
+            LogPrint(BCLog::NET, "disconnecting due to inactivity peer=%d\n", pnode->GetId());
```
`InactivityCheck()` already logs a reason for any case where it returns true, eg:

```
2023-09-26T07:21:14.013289Z [net] socket receive timeout: 1201s peer=745   <==== this one
2023-09-26T07:21:14.014945Z [net] disconnecting peer=745
2023-09-26T07:21:14.016616Z [net] Cleared nodestate for peer=745

2023-09-26T08:33:15.016157Z [net] version handshake timeout peer=9862    <=== and this one
2023-09-26T08:33:15.016539Z [net] disconnecting peer=9862
2023-09-26T08:33:15.017321Z [net] Cleared nodestate for peer=9862
```

Does an extra log entry here actually help? Would adding the word "disconnecting" to the log entries in `InactivityCheck()` be better?

Sjors replied:
Done, and also made sure the keyword `disconnecting` appears in all cases where we disconnect.

RandyMcMillan commented:
ACK 61a208a1d8d970558ffcf582b5573f3f3219caa9

ajtowns commented on src/net.cpp:
```diff
@@ -2108,6 +2108,7 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
             {
                 bool notify = false;
                 if (!pnode->ReceiveMsgBytes({pchBuf, (size_t)nBytes}, notify)) {
+                    LogPrint(BCLog::NET, "receiving message bytes failed for peer=%d\n", pnode->GetId());
```
This one didn't get "disconnecting" added?

Sjors replied:
No, because `CloseSocketDisconnect` already uses that term. I updated the PR description (yesterday) to explain the reasoning here.

hodlinator replied:
(This is currently being resolved by adding a call to `DisconnectMsg()` unless I've missed something).

https://github.com/bitcoin/bitcoin/blob/06443b8f28bcec4315cec262eb03343dee5465a6/src/net.cpp#L2146-L2153

Sjors commented:
Rebased.

Sjors commented:
I added `peeraddr=` to each of them. This is indeed quite useful, because we can't look up the IP after disconnection. Adding `net=` also makes sense, but let's do that everywhere in another PR (probably using a helper function).

Sjors commented:
@maflcko indeed it is, though for inbound it depends on the exact log settings. I think right now you won't see these disconnect messages without `-debug=net` and with it you'll see all the connect messages. But if we increase the relative priority of disconnect messages then you might not. So it still seems useful to add.

sr-gi commented:
Concept ACK. Just a single consideration: wouldn't it also make sense to test this with `-logips=1`?

Sjors commented:
@sr-gi I could, but I prefer to punt that until we have helper function that can be tested in just one place: https://github.com/bitcoin/bitcoin/pull/28521#issuecomment-1835692974

davidgumberg commented:
I may be mistaken in understanding you to be saying something else, but it seems to me that any time `pnode->fDisconnect` is set to `true`, this will result in a call to `CloseSocketDisconnect()` and a disconnect log will happen.

Maybe this doesn't affect the purpose of this PR, as the connection close may come some time later than `pnode->fDisconnect` and using 'disconnect' consistently makes it easier to connect the message from `CloseSocketDisconnect()` with the reason that is logged above.

In [`CConnman::DisconnectNodes()`](https://github.com/bitcoin/bitcoin/blob/eb0bdbdd753bca97120247b921fd29d606fea6e9/src/net.cpp#L1913) :

```cpp
void CConnman::DisconnectNodes()
{
    // [...]
    {
        // [...]
        // Disconnect unused nodes
        std::vector<CNode*> nodes_copy = m_nodes;
        for (CNode* pnode : nodes_copy)
        {
            if (pnode->fDisconnect)
            {
                // remove from m_nodes
                m_nodes.erase(remove(m_nodes.begin(), m_nodes.end(), pnode), m_nodes.end());
                // [...]
                // close socket and cleanup
                pnode->CloseSocketDisconnect();
                // [...]
            }
        }
    }
    // [...]
}
```

For example, I tested setting `fNetworkActive` with the `setnetworkactive` rpc and got the following logs on this branch:
```console
2024-05-04T19:06:56Z SetNetworkActive: false
2024-05-04T19:06:56Z [net] Network not active, disconnecting peer=0
2024-05-04T19:06:56Z [net] Network not active, disconnecting peer=1
2024-05-04T19:06:56Z [net] Network not active, disconnecting peer=2
2024-05-04T19:06:56Z [net] disconnecting peer=0
2024-05-04T19:06:56Z [net] disconnecting peer=1
2024-05-04T19:06:56Z [net] disconnecting peer=2
```

This should probably also use the #28318 `LogDebug` interface over `LogPrint`.

Sjors commented:
Rebased after #29431

@davidgumberg this PR makes it easier to do something like `cat ~/.bitcoin/debug.log | grep disconnect`, and avoid the need for e.g. `grep -n2` (or more) to get additional context.

I switched to `LogDebug` in the lines touched by this PR.

Sjors commented:
Rebased after #30420.

davidgumberg commented:
Re-reviewed and manually tested ACK https://github.com/bitcoin/bitcoin/commit/45958d595b70222c89c25768f357a2ebbafac2b1

It should be easy to grep for the reason that a peer got disconnected, and logging the peer's address when `fLogIPs` seems helpful.

Tested by running:

```bash
./bitcoind -datadir=/tmp/trash -debug=net -logips
```

made `fNetworkActive == false` by:

```bash
./bitcoin-cli -datadir=/tmp/trash setnetworkactive false
```

<details>

<summary>Got the following output:</summary>

```
2024-07-23T01:08:08Z SetNetworkActive: false
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=0 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=1 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=2 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=4 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=5 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] Network not active, disconnecting peer=6 peeraddr=[REDACTEDIP]:8333
2024-07-23T01:08:08Z [net] disconnecting peer=0
2024-07-23T01:08:08Z [net] disconnecting peer=1
2024-07-23T01:08:08Z [net] disconnecting peer=2
2024-07-23T01:08:08Z [net] disconnecting peer=4
2024-07-23T01:08:08Z [net] disconnecting peer=5
2024-07-23T01:08:08Z [net] disconnecting peer=6
2024-07-23T01:08:08Z [net] disconnecting peer=8
2024-07-23T01:08:08Z [net] disconnecting peer=10
2024-07-23T01:08:08Z [net] disconnecting peer=11
```

</details>

Micro-nit: if the purpose is to reduce the need to do `grep -n2` , then it seems to me we should add the word 'disconnect' to *all* [instances](https://github.com/bitcoin/bitcoin/blob/8754d055c65e11fd2afa59f9e5de7c60a9e0ec23/src/net.cpp#L1634) [where](https://github.com/bitcoin/bitcoin/blob/45958d595b70222c89c25768f357a2ebbafac2b1/src/net.cpp#L2144) the reason appears in the line above, and *if* we were to always log the reason for disconnect when it happens,  it seems to me the extra generic log in `CloseSocketDisconnect` should be dropped.

davidgumberg commented on src/net.cpp:
```diff
@@ -1968,25 +1972,43 @@ bool CConnman::InactivityCheck(const CNode& node) const
     }
 
     if (!node.fSuccessfullyConnected) {
+        const std::string disconnect_msg{strprintf(", disconnecting peer=%d%s",
```
nit: Maybe this should move to the top of the function and be used in the other three disconnect logs here
```diff
 if (!ShouldRunInactivityChecks(node, now)) return false;
+const std::string disconnect_msg{...};
```

Sjors committed:
net: additional disconnection logging

Use the word "disconnecting" everywhere for easier grep.

Sjors committed:
net_processing: use CNode::DisconnectMsg helper

This is not a pure refactor:
1. It slightly changes the log messages, as reflected in the test changes
2. It adds the IP address to all disconnect logging (when fLogIPs is set)

Sjors committed:
net: add LogIP() helper, use in net_processing

Sjors commented:
Thanks @davidgumberg. I added a helper function `CNode::DisconnectMsg`, and it's now also used for:
* "receiving message bytes failed"
* "socket send error for"
* "socket recv error"
* "socket closed"

In order to get rid of the redundant generic "disconnecting peer" in `CNode::CloseSocketDisconnect()`, I added the log statement to the two remaining call sites.

Since it's a method on `CNode` I can use it in net_processing as well, for which I added a separate commit. It has the effect of slightly modifying some of the existing log messages, and adds IP address logging in a few more places.

I did not apply this to:
* "connected to self"
* some messages that would require a more complicated helper:
  * "disconnecting but not discouraging" 
  * "disconnecting extra block-relay-only peer=%d"
  * "disconnecting extra outbound peer"
  
While I was on a roll, I also added a `LogIP()` helper (as suggested earlier). There's only a small number of places where we print the IP address that's not in the context of a disconnect - but now it's easier to expand that. I skipped the "send version message" message log here.

Sjors commented:
Pushed an update to consistently replace any `LogPrint` we touch with `LogDebug`, and every `LogPrintf ` with `LogInfo`. Also fixed some missing `\n` characters as found by tidy.

davidgumberg commented on src/net.cpp:
```diff
@@ -690,11 +690,16 @@ bool CNode::ReceiveMsgBytes(Span<const uint8_t> msg_bytes, bool& complete)
     return true;
 }
 
+std::string CNode::LogIP(bool log_ip) const
+{
+    return log_ip ? strprintf(" peeraddr=%s", addr.ToStringAddrPort()) : "";
+}
+
```
Why not:
```suggestion
std::string CNode::LogIP() const
{
    return fLogIPs ? strprintf(" peeraddr=%s", addr.ToStringAddrPort()) : "";
}
```

Sjors replied:
I think we should avoid accessing the global variable `fLogIPs`. A future refactor could add `m_log_ips` to `PeerManagerImpl` and/or `CConnman`, initialize it from `gArgs` and then drop `fLogIPs`.

Sjors commented:
Rebased after #30750. It made the diff slightly smaller, since I had made the same changes on every line touched here.

naumenkogs commented:
ACK 08cd29f5020995294a4f6274e72afab58bf69aa2

The first two commits make logging cleaner and more uniform.
I'm not sure the last commit is useful, but it doesn't hurt either.

davidgumberg commented on src/net_processing.cpp:
```diff
@@ -6240,7 +6243,7 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
             // the download window should be much larger than the to-be-downloaded set of blocks, so disconnection
             // should only happen during initial block download.
-            LogPrintf("Peer=%d%s is stalling block download, disconnecting\n", pto->GetId(), fLogIPs ? strprintf(" peeraddr=%s", pto->addr.ToStringAddrPort()) : "");
+            LogInfo("Peer is stalling block download%s\n", pto->DisconnectMsg(fLogIPs));
```
nit:
```suggestion
            LogInfo("Peer is stalling block download, %s\n", pto->DisconnectMsg(fLogIPs));
```

Sjors replied:
Fixed.

davidgumberg commented on src/net.cpp:
```diff
@@ -1902,6 +1908,7 @@ void CConnman::DisconnectNodes()
                 pnode->grantOutbound.Release();
 
                 // close socket and cleanup
+                LogDebug(BCLog::NET, "%s\n", pnode->DisconnectMsg(fLogIPs));
```
I believe your PR makes sure that `DisconnectMsg` gets printed any time `fDisconnect == true` so this line results in a duplicate log in each of those instances.

For example when doing `bitcoin-cli setnetworkactive false`:

```console
2024-09-18T00:32:49Z SetNetworkActive: false
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=0 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=1 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=3 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=4 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=5 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=6 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] Network not active, disconnecting peer=7 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=0 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=1 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=3 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=4 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=5 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=6 peeraddr=xx.xx.xx.xx:8333
2024-09-18T00:32:49Z [net] disconnecting peer=7 peeraddr=xx.xx.xx.xx:8333
```

Sjors replied:
You're right. I dropped this line and checked again that anytime `fDisconnect` is set to `true` emits a log message.

Sjors commented:
Rebased and dropped duplicate message in `CConnman::DisconnectNodes`.

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/9967e8feb62a0c30612e437b70d1d64a2e67b9e0

Thanks for putting up with the review churn on this, it looks great! This PR makes peer disconnect message logging more consistent, and makes us better respect the `-logips` argument.

I reviewed the range-diff:

```bash
git range-diff a74bdeea..08cd29f5 6fc46927..9967e8fe
```

<details>

<summary>Manual testing</summary>

Starting a node and then doing `bitcoin-cli setnetworkactive false` and got the following:

```console
2024-09-20T17:47:15Z SetNetworkActive: false
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=0
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=2
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=4
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=5
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=6
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=8
2024-09-20T17:47:15Z [net] Network not active, disconnecting peer=9
```

Here is the disconnect logging that happens when a socket closes:

```console
2024-09-20T17:58:09Z [net] Added connection peer=1
2024-09-20T17:58:09Z [net] start sending v2 handshake to peer=1
2024-09-20T17:58:09Z [net] sending version (103 bytes) peer=1
2024-09-20T17:58:09Z [net] send version message: version 70016, blocks=0, txrelay=1, peer=1
2024-09-20T17:58:09Z [net] socket closed, disconnecting peer=1
2024-09-20T17:58:09Z [net] retrying with v1 transport protocol for peer=1
2024-09-20T17:58:09Z [net] Cleared nodestate for peer=1
2024-09-20T17:58:09Z [net] trying v1 connection xx.xx.xx.xx:8333 lastseen=27.2hrs
```

invalid transaction in an INV:

```console
2024-09-20T18:01:10Z [net] received: pong (8 bytes) peer=14
2024-09-20T18:01:10Z [net] received: inv (253 bytes) peer=14
2024-09-20T18:01:10Z [net] transaction (d9db66309820a33f00b0cc8c843cd8683fdc03357dbe1959a9de18afc04ff675) i
nv sent in violation of protocol, disconnecting peer=14
2024-09-20T18:01:10Z [net] Cleared nodestate for peer=14
```

service flags:

```console
2024-09-20T18:01:17Z [net] Added connection peer=15
2024-09-20T18:01:17Z [net] sending version (103 bytes) peer=15
2024-09-20T18:01:17Z [net] send version message: version 70016, blocks=0, txrelay=0, peer=15
2024-09-20T18:01:17Z [net] received: version (102 bytes) peer=15
2024-09-20T18:01:17Z [net] peer does not offer the expected services (0000040c offered, 00000009 expected),
 disconnecting peer=15
2024-09-20T18:01:17Z [net] Cleared nodestate for peer=15
```

shutdown:

```console
2024-09-20T18:03:00Z [net] disconnecting peer=0
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=0
2024-09-20T18:03:00Z [net] disconnecting peer=3
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=3
2024-09-20T18:03:00Z [net] disconnecting peer=4
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=4
2024-09-20T18:03:00Z [net] disconnecting peer=5
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=5
2024-09-20T18:03:00Z [net] disconnecting peer=7
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=7
2024-09-20T18:03:00Z [net] disconnecting peer=8
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=8
2024-09-20T18:03:00Z [net] disconnecting peer=9
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=9
2024-09-20T18:03:00Z [net] disconnecting peer=10
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=10
2024-09-20T18:03:00Z [net] disconnecting peer=13
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=13
2024-09-20T18:03:00Z [net] disconnecting peer=17
2024-09-20T18:03:00Z [net] Cleared nodestate for peer=17
```

</details>

vasild commented on src/net.cpp:
```diff
@@ -1967,25 +1978,37 @@ bool CConnman::InactivityCheck(const CNode& node) const
 
     if (now > last_recv + TIMEOUT_INTERVAL) {
-        LogDebug(BCLog::NET, "socket receive timeout: %is peer=%d\n", count_seconds(now - last_recv), node.GetId());
+        LogDebug(BCLog::NET,
+            "socket receive timeout: %is%, s\n", count_seconds(now - last_recv),
+            node.DisconnectMsg(fLogIPs)
```
```suggestion
        LogDebug(BCLog::NET,
            "socket receive timeout: %is, %s\n", count_seconds(now - last_recv),
            node.DisconnectMsg(fLogIPs)
```

Sjors replied:
Fixed

vasild commented on src/net.cpp:
```diff
@@ -1967,25 +1978,37 @@ bool CConnman::InactivityCheck(const CNode& node) const
 
     if (last_recv.count() == 0 || last_send.count() == 0) {
-        LogDebug(BCLog::NET, "socket no message in first %i seconds, %d %d peer=%d\n", count_seconds(m_peer_connect_timeout), last_recv.count() != 0, last_send.count() != 0, node.GetId());
+        LogDebug(BCLog::NET,
+            "socket no message in first %i seconds, %d %d, %s\n",
+            count_seconds(m_peer_connect_timeout),
+            last_recv.count() != 0,
+            last_send.count() != 0,
+            node.DisconnectMsg(fLogIPs)
```
This `%d %d`, `last_recv.count() != 0, last_send.count() != 0` is super obscure. It was like that before, so it is fine wrt this PR.

But would be nice to fix it, since this is changing the same log message already. I am not sure how though. Maybe change that `%d` to something like "has never received from the peer" + "has never sent to the peer".

vasild commented on src/net_processing.cpp:
```diff
@@ -4094,24 +4093,24 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
     }
 
     if (!pfrom.fSuccessfullyConnected) {
-        LogDebug(BCLog::NET, "Unsupported message \"%s\" prior to verack from peer=%d\n", SanitizeString(msg_type), pfrom.GetId());
+        LogDebug(BCLog::NET, "Unsupported message \"%s\" prior to verack, %s\n", SanitizeString(msg_type), pfrom.DisconnectMsg(fLogIPs));
         return;
     }
```
This log shouldn't contain "disconnecting" because we do not disconnect.

Sjors replied:
Reverted this change (and its corresponding test).

vasild commented on src/net.cpp:
```diff
@@ -1874,7 +1885,7 @@ void CConnman::DisconnectNodes()
             // Disconnect any connected nodes
```
_(suggestion at random place in the code)_

Consider to also adjust the `CConnman::DisconnectNode()` functions.

Sjors replied:
Tempted to leave the PR as is for now as it's already quite large.

vasild reviewed (commented):
Approach ACK 9967e8feb6

I noticed the repeating pattern `fLogIPs ? strprintf(", peeraddr=%s", pfrom.addr.ToStringAddrPort()) : ""` while working on https://github.com/bitcoin/bitcoin/pull/29415 and I repeated it a lot of times in the new messages I added :disappointed:. It has been on my TODO to introduce some function to deduplicate that. Thanks for the `CNode::LogIP()` method!

naumenkogs commented:
Would be great to address @vasild feedbacks :)

Sjors commented:
Seems like I need to look into and probably fix at least https://github.com/bitcoin/bitcoin/pull/28521#discussion_r1791784114

Will look into the other comments too.

Sjors commented:
Rebased and addressed feedback.

Sjors commented:
Fixed [tidy complaint](https://cirrus-ci.com/task/4855637697363968?logs=ci#L3329-L3332).

Sjors committed:
net: clarify if we ever sent or received from peer

vasild reviewed (approved):
ACK 06443b8f28bcec4315cec262eb03343dee5465a6

Sjors commented:
@0xB10C @naumenkogs @davidgumberg one more review round pls?

danielabrozzoni reviewed (approved):
ACK 06443b8f28bcec4315cec262eb03343dee5465a6 

Thanks for working on this! I've found myself looking through logs and struggling to understand why some connections were closed, this will come in handy :)

davidgumberg commented:
reACK https://github.com/bitcoin/bitcoin/commit/06443b8f28bcec4315cec262eb03343dee5465a6

Retested using the same procedure as [above](https://github.com/bitcoin/bitcoin/pull/28521#issuecomment-2364259064), `InactivityCheck()` logging is now much less opaque, and this PR makes disconnect logging consistent and better respect the `-logips` argument

hodlinator reviewed (commented):
Concept ACK 06443b8f28bcec4315cec262eb03343dee5465a6

Having done a similar change to facilitate disconnection issue debugging in a different networked system, I fully support this initiative.

Excellent approach to enforce uniform wording through a log helper function.

An alternative but more disruptive approach would have been to make `CNode::fDisconnect` private and add `CNode::Disconnect(std::string_view reason)`.

Sorry I didn't see this PR earlier. I completely understand if you'd rather defer further changes until other PRs.

Sjors commented:
@hodlinator thanks for the review. I'd rather not expand the scope of this PR even more, because it's painful to keep rebasing (as I increasingly forget what I did in the first place).

If I need to make other more substantial changes, I'll look into adding a few more log statements based on your feedback, but otherwise let's leave it to a followup.

hodlinator reviewed (approved):
ACK 06443b8f28bcec4315cec262eb03343dee5465a6

Agree that follow-ups should be handled in other PRs at this point.

---

<details><summary>
Tested syncing signet and turned off WiFi
</summary>

```
2024-12-20T21:34:36Z [net] got inv: wtx 3e9f4e49ddfb5cfd1fd847de22836de941a7245054db0629bb5c0cdfdd2466bf  have peer=0
2024-12-20T21:34:36Z [net] received: tx (248 bytes) peer=7
2024-12-20T21:34:37Z [net] sending inv (73 bytes) peer=5
2024-12-20T21:34:37Z [net] sending inv (37 bytes) peer=4
2024-12-20T21:34:37Z [net] sending inv (73 bytes) peer=3
2024-12-20T21:34:37Z [net] sending inv (73 bytes) peer=1
2024-12-20T21:34:37Z [net] sending inv (73 bytes) peer=6
2024-12-20T21:34:38Z [net] sending ping (8 bytes) peer=4
2024-12-20T21:34:38Z [net] sending inv (37 bytes) peer=0
2024-12-20T21:34:41Z [net] sending inv (73 bytes) peer=2
2024-12-20T21:35:03Z [net] sending addrv2 (16 bytes) peer=4
2024-12-20T21:35:19Z [net] sending addrv2 (28 bytes) peer=5
2024-12-20T21:35:22Z [net] sending addrv2 (28 bytes) peer=7
2024-12-20T21:35:28Z [net] sending ping (8 bytes) peer=5
2024-12-20T21:36:17Z [net] sending ping (8 bytes) peer=6
2024-12-20T21:36:17Z [net] sending ping (8 bytes) peer=7
2024-12-20T21:36:18Z [net] sending ping (8 bytes) peer=0
2024-12-20T21:36:23Z [net] sending ping (8 bytes) peer=8
2024-12-20T21:36:24Z [net] sending ping (8 bytes) peer=9
2024-12-20T21:36:25Z [net] sending ping (8 bytes) peer=1
2024-12-20T21:36:31Z [net] sending ping (8 bytes) peer=2
2024-12-20T21:36:32Z [net] sending ping (8 bytes) peer=3
2024-12-20T21:38:53Z [net] Making feeler connection to 18.190.155.122:38333
2024-12-20T21:38:53Z [net] trying v1 connection 18.190.155.122:38333 lastseen=12.6hrs
2024-12-20T21:38:53Z [net] connect() to 18.190.155.122:38333 failed: Network is unreachable (101)
2024-12-20T21:40:27Z [net] Making feeler connection to 102.97.13.193:38333
...
2024-12-20T21:50:53Z [net] socket recv error, disconnecting peer=0: Connection timed out (110)
...
2024-12-20T21:51:05Z [net] socket recv error, disconnecting peer=4: Connection timed out (110)
...
2024-12-20T21:51:18Z [net] socket recv error, disconnecting peer=1: Network is unreachable (101)
...
2024-12-20T21:51:22Z [net] socket recv error, disconnecting peer=2: Connection timed out (110)
...
2024-12-20T21:52:48Z [net] socket recv error, disconnecting peer=9: Connection timed out (110)
...
2024-12-20T21:52:52Z [net] socket recv error, disconnecting peer=8: Network is unreachable (101)
...
2024-12-20T21:52:56Z [net] socket recv error, disconnecting peer=5: Connection timed out (110)
...
2024-12-20T21:53:00Z [net] socket recv error, disconnecting peer=3: Connection timed out (110)
...
2024-12-20T21:53:21Z [net] socket recv error, disconnecting peer=6: Connection timed out (110)
...
2024-12-20T21:53:49Z [net] socket recv error, disconnecting peer=7: Connection timed out (110)
```

</details>

---

Maybe the PR summary could be touched up slightly:

> All cases where we disconnect now come with a log message that has the word `disconnecting`:
>
>* `CloseSocketDisconnect()` logs `disconnecting peer=…`

This is describing the state on master I guess, but all the other points describe what is changed on this PR.

>* whereever we set `pnode->fDisconnect = true;`

Except 2 cases, a & b here: https://github.com/bitcoin/bitcoin/pull/28521#discussion_r1890824361

Typo: Should be [wherever](https://english.stackexchange.com/questions/198261/why-is-it-wherever-instead-of-whereever)

Sjors commented:
Fixed the typo in the description. 

> This is describing the state on master I guess, but all the other points describe what is changed on this PR.

I added "all calls to" to the first item.

Added a link to your comment for the list of exceptions.
